### PR TITLE
ensure the cleaning of overshadowed unloaded segments

### PR DESCRIPTION
**Pull request by b-slim**
Druid will try infinitely try to load segments that are unloaded even if the segment is actually overshadowed by a loaded segment.
We had this issue, where a set of corrupted segments were in the load queue forever even we have re indexed the data and published/loaded new segments.
This patch will drop all segments that are not loaded yet and overshadowed by newer version.

### CODE REVIEW

**Comment by fjy** [comment]
Oh sweet, this bug is finally getting fixed :P

**Comment by drcrallen** [comment]
Is this a documented bug?

**Comment by b-slim** [comment]
@drcrallen i am not use, if you are asking if there is an existing documentation or if i have to add this documentation ?

**Comment by drcrallen** [comment]
@b-slim I mean @fjy referred to something finally getting fixed, and I was curious if this was a something that already had an issue.

**Comment by drcrallen** [comment]
@b-slim can you add tests for this?

File: server/src/main/java/io/druid/server/coordinator/helper/DruidCoordinatorCleanupOvershadowed.java [hunk 1]
```diff
@@ -82,6 +83,19 @@ public DruidCoordinatorRuntimeParams run(DruidCoordinatorRuntimeParams params)
             stats.addToGlobalStat("overShadowedCount", 1);
           }
         }
+
+        for (LoadQueuePeon loadQueue : coordinator.getLoadManagementPeons().values()) {
```
Comment (drcrallen) [inline server/src/main/java/io/druid/server/coordinator/helper/DruidCoordinatorCleanupOvershadowed.java]: Can you simply add the loadQueues to the timeline before the immediately prior loop, and let the loop which looks for overshadowed segments find them?

**Comment by fjy** [comment]
@drcrallen this bug has been around from a time where I think I was the only person in the world working on Druid

File: server/src/main/java/io/druid/server/coordinator/helper/DruidCoordinatorCleanupOvershadowed.java (hunk 1 above)
Comment (b-slim) [inline server/src/main/java/io/druid/server/coordinator/helper/DruidCoordinatorCleanupOvershadowed.java]: @drcrallen you can not, you want to make sure that what is overshadowing is loaded.

Comment (drcrallen) [inline server/src/main/java/io/druid/server/coordinator/helper/DruidCoordinatorCleanupOvershadowed.java]: I see it now thanks.

File: server/src/main/java/io/druid/server/coordinator/helper/DruidCoordinatorCleanupOvershadowed.java [hunk 2]
```diff
@@ -82,6 +83,19 @@ public DruidCoordinatorRuntimeParams run(DruidCoordinatorRuntimeParams params)
...
+          for (DataSegment dataSegment: loadQueue.getSegmentsToLoad()) {
+            timeline = timelines.get(dataSegment.getDataSource());
+            if (timeline == null) continue;
+            timeline.add(dataSegment.getInterval(), dataSegment.getVersion(), dataSegment.getShardSpec().createChunk(dataSegment));
```
Comment (drcrallen) [inline server/src/main/java/io/druid/server/coordinator/helper/DruidCoordinatorCleanupOvershadowed.java]: can you add a comment here explaining why this is in its own loop? something like 

```java
// Temporarily add queued segments to the timeline to see if they still need to be loaded.
```

**Comment by b-slim** [comment]
@drcrallen added tests and comments.

**Comment by drcrallen** [comment]
Sweet 👍

**Comment by nishantmonu51** [comment]
:+1: , awesome, great fix. 
@drcrallen @fjy Here is the related issue - https://github.com/druid-io/druid/issues/540

**Comment by nishantmonu51** [comment]
travis failure - https://travis-ci.org/druid-io/druid/jobs/134604151

**Comment by nishantmonu51** [comment]
fixed #540

**Comment by drcrallen** [comment]
@b-slim if this is for 0.9.1 it needs a backport

**Pull request merged**